src/join/composer.jl

```diff
@@ -182,11 +182,56 @@ function _propagate_join_metadata!(joiner::DataFrameJoiner, dfr_noon::AbstractDa
     return nothing
 end
 
+function _count_sortperm(input::Vector{Int})
+    isempty(input) && return UInt32[]
+    vmin, vmax = extrema(input)
+    delta = vmin - 1
+    Tc = vmax - delta < typemax(UInt32) ? UInt32 : Int
+    Tp = length(input) < typemax(UInt32) ? UInt32 : Int
+    _count_sortperm(input, zeros(Tc, vmax - delta + 1),
+                    Vector{Tp}(undef, length(input)), delta)
+end
+
+function _count_sortperm(input::Vector{Int}, count::Vector,
+                         output::Vector, delta::Int)
+    @assert firstindex(input) == 1
+    for j in input
+        count[j - delta] += 1
+    end
```

Comment:
All three loops can be theoretically `@inbounds`, however, since the algorithm is non-trivial I think it is better to leave it as is for now. The impact on performance is low.
I added a comment that `@inbounds` can be added in the future.